In [1]:
import pandas as pd
import numpy as np

In [4]:
df=pd.read_csv('open_pubs.csv',header=None, keep_default_na=False)
df.head()

0                   1                                                  2  \
0  22          Anchor Inn        Upper Street, Stratford St Mary, COLCHESTER   
1  36  Ark Bar Restaurant  Ark Bar And Restaurant, Cattawade Street, Bran...   
2  74           Black Boy  The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...   
3  75         Black Horse        Lower Street, Stratford St Mary, COLCHESTER   
4  76          Black Lion                      Lion Road, Glemsford, SUDBURY   

          3       4        5          6         7        8  
0   CO7 6LW  604749   234404  51.970379  0.979340  Babergh  
1  CO11 1RH  610194   233329  51.958698  1.057832  Babergh  
2  CO10 2EA  587334   241316  52.038595  0.729915  Babergh  
3   CO7 6JS  622675 -5527598         \N        \N  Babergh  
4  CO10 7RF  622675 -5527598         \N        \N  Babergh

In [5]:
df.columns=['fsa_id','name','address','postcode','easting','northing','latitude','longitude','local_authority']
df.head()

fsa_id                name  \
0      22          Anchor Inn   
1      36  Ark Bar Restaurant   
2      74           Black Boy   
3      75         Black Horse   
4      76          Black Lion   

                                             address  postcode  easting  \
0        Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1  Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2  The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
3        Lower Street, Stratford St Mary, COLCHESTER   CO7 6JS   622675   
4                      Lion Road, Glemsford, SUDBURY  CO10 7RF   622675   

   northing   latitude longitude local_authority  
0    234404  51.970379  0.979340         Babergh  
1    233329  51.958698  1.057832         Babergh  
2    241316  52.038595  0.729915         Babergh  
3  -5527598         \N        \N         Babergh  
4  -5527598         \N        \N         Babergh

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51331 entries, 0 to 51330
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   fsa_id           51331 non-null  int64 
 1   name             51331 non-null  object
 2   address          51331 non-null  object
 3   postcode         51331 non-null  object
 4   easting          51331 non-null  int64 
 5   northing         51331 non-null  int64 
 6   latitude         51331 non-null  object
 7   longitude        51331 non-null  object
 8   local_authority  51331 non-null  object
dtypes: int64(3), object(6)
memory usage: 3.5+ MB


In [7]:
df.isna().sum()


fsa_id             0
name               0
address            0
postcode           0
easting            0
northing           0
latitude           0
longitude          0
local_authority    0
dtype: int64

##### No Null values but \N values are present . That need to be cleaned.


### Replacing \N with NaN

In [8]:
def change_N_to_nan(col_val):
    if col_val==r'\N':
        return np.nan
    else:
        return col_val

df['fsa_id']=df['fsa_id'].apply(change_N_to_nan)
df['name']=df['name'].apply(change_N_to_nan)
df['address']=df['address'].apply(change_N_to_nan)
df['postcode']=df['postcode'].apply(change_N_to_nan)
df['easting']=df['easting'].apply(change_N_to_nan)
df['northing']=df['northing'].apply(change_N_to_nan)
df['latitude']=df['latitude'].apply(change_N_to_nan)
df['longitude']=df['longitude'].apply(change_N_to_nan)
df['local_authority']=df['local_authority'].apply(change_N_to_nan)

In [9]:
df.isna().sum()

fsa_id               0
name                 0
address              0
postcode             0
easting              0
northing             0
latitude           767
longitude          767
local_authority      0
dtype: int64

##### these is 767 values of NaN in Latitude as well as Longitude. So wee need to drop those values.

In [10]:
df.dropna(axis=0, inplace=True)
df.isna().sum()

fsa_id             0
name               0
address            0
postcode           0
easting            0
northing           0
latitude           0
longitude          0
local_authority    0
dtype: int64

##### Latitude and Longitude must be in numeric form.

In [11]:
df['longitude']=df['longitude'].astype(float)
df['latitude']=df['latitude'].astype(float)

In [12]:
df.dtypes

fsa_id               int64
name                object
address             object
postcode            object
easting              int64
northing             int64
latitude           float64
longitude          float64
local_authority     object
dtype: object

In [13]:
df['name'].nunique()

35809

##### total number of pubs are 35809

In [14]:
df['local_authority'].nunique()

360

##### total number of local authority are 360

In [18]:
df.to_csv('open_pubs_clean.csv', index=False)

In [20]:
df.head()

fsa_id                name  \
0      22          Anchor Inn   
1      36  Ark Bar Restaurant   
2      74           Black Boy   
5      97        Brewers Arms   
6     102        Bristol Arms   

                                             address  postcode  easting  \
0        Upper Street, Stratford St Mary, COLCHESTER   CO7 6LW   604749   
1  Ark Bar And Restaurant, Cattawade Street, Bran...  CO11 1RH   610194   
2  The Lady Elizabeth, 7 Market Hill, SUDBURY, Su...  CO10 2EA   587334   
5  The Brewers Arms, Bower House Tye, Polstead, C...   CO6 5BZ   598743   
6                     Bristol Hill, Shotley, IPSWICH   IP9 1PU   624624   

   northing   latitude  longitude local_authority  
0    234404  51.970379   0.979340         Babergh  
1    233329  51.958698   1.057832         Babergh  
2    241316  52.038595   0.729915         Babergh  
5    240655  52.028694   0.895650         Babergh  
6    233550  51.955042   1.267642         Babergh